### DISCUSSION TITLE
Alternate scenarios evenly instead of randomly selecting based on weights

### DISCUSSION THREAD

**Discussion by SnowdenWintermute** [artilleryio/artillery#1837] [Artillery]
I have a Socket.IO app in which users can host and join each other's games. I want to make a load test in which half of the VU's host games and the other half join the games, then they both emit some game input events. I figure I could have a scenario for a VU hosting a game and one for a VU joining a game.

Host scenario: 
- host a game
- think for a few seconds so the other scenario's VU has time to join the game
- have the game start automatically once two users are in the game (my app handles this)
- emit some user inputs for the game

In the second scenario the VUs would:
- join a game
- think for a few seconds while the game starts
- emit some user inputs for the game

For this to work I would want the two scenarios to be chosen evenly and alternate, with the first VU created hosting a game and the second VU joining it.

I also need some way to store a variable to be used as the game name so the VU's can specify what game to join. This variable would increment every time the first scenario runs, and also be accessible to the second scenario so that VU can join the correct game.

Is this possible?

**Comment by hassy** [comment]
hi @SnowdenWintermute 👋 

Even distribution of scenarios - yep, that's the default behavior when weights aren't assigned explicitly. If you run a test script with two scenarios, Artillery will assign equal weights to them.

Sharing a variable between two VUs - that's not possible without some form of external state storage. VUs are meant to mimic real-world clients which don't have shared memory, unless they both access some sort of external storage. Depending on where your test is running a Redis instance or the local file system could work for that.

**Reply by SnowdenWintermute** [reply]
If equal weights are assigned do the scenarios alternate, or are they randomly selected with a 50/50 chance with each new VU?

**Reply by hassy** [reply]
Randomly selected with 50/50 chance.

**Reply by SnowdenWintermute** [reply]
If alternating scenarios in a consistent manner is not possible, is there another approach for getting two VUs to interact with each other in a predictable way?

One idea I had was having a single scenario which accesses a variable from a JavaScript function which in turn accesses Redis and increments the variable each time it is called. That way we could at least know if we were on an even or odd instance of the scenario and then have the VU play the role of game host or game joiner based on that.

This would require some kind of if/else statements in the scenario like the ifTrue feature described in https://www.artillery.io/docs/guides/guides/http-reference#examples-1, but can ifTrue be used with "- emit" or only with http requests? I tried it out and it didn't seem to have any effect.

So another thought was to again go with two scenarios, one for the game host and one for the joiner and settle for the fact that it is random how many we would get of each. The VUs would know what game name to host or join based on a variable accessed from a JavaScript function which stores in Redis the name of the last game hosted and increments it each time the function is called by the game host scenario. This has the unfortunate effect that some VUs would be created as game joiners before a VU hosted a game for them to join, or that the last few VU would be hosting a game and have no-one to join it.

Both my ideas seem hacky and roundabout so I'm wondering if there is a more straightforward approach that I'm missing.
